### ISSUE TITLE
Memory usage increases when webp is used after 5.13.0

### ISSUE THREAD

**Issue by kirou** [SDWebImage/SDWebImage#3417]
### New Issue Checklist

* [ x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios 
 Platform Version        | 14.0 〜
 SDWebImage Version      | 5.13.0 〜
 Integration Method      | carthage
 Xcode Version           | Xcode 14
 Repro rate              | all the time (100%) 

### Issue Description and Steps

I am using SDWebImage/SDWebImageWebPCoder to display webp images.
After upgrading OSS, the memory usage increases as below.

- before
<img width="254" alt="スクリーンショット 2022-10-21 10 36 37" src="https://user-images.githubusercontent.com/1872033/197093944-917575cb-3c2c-42ac-a80a-35c4509b28f0.png">

- after
<img width="254" alt="スクリーンショット 2022-10-21 9 58 00" src="https://user-images.githubusercontent.com/1872033/197093967-7e5b641a-e691-4540-95ef-a4cd2fa45b20.png">

- changed version
```
SDWebImage: 5.12.6 -> 5.13.4 (This phenomenon occurs from 5.13.0)
SDWebImageWebPCoder: 0.8.5 -> 0.9.1
libwebp-Xcode: 1.2.1
```

If you quit webp mode and display in jpg, the memory will not increase.

```
SDImageCodersManager.shared.addCoder(SDImageWebPCoder.shared)
↓
//SDImageCodersManager.shared.addCoder(SDImageWebPCoder.shared)
```

Is there a way around this?
I've read the documentation, sorry if I've overlooked something.

**Comment by zhouyantongiosDev** [comment]
Same question

**Comment by dreampiggy** [comment]
I think this **Visible RAM increasement** is not what actually increase. WebP coder does not has any lazy-decoding issue.

Use instruments's Allocation template, not this Xcode memory hint tool.

For more defail, read my blog post carefully about **Force Decode/Lazy Decoding** related thing.

https://dreampiggy.com/2019/01/18/%E4%B8%BB%E6%B5%81%E5%9B%BE%E7%89%87%E5%8A%A0%E8%BD%BD%E5%BA%93%E6%89%80%E4%BD%BF%E7%94%A8%E7%9A%84%E9%A2%84%E8%A7%A3%E7%A0%81%E7%A9%B6%E7%AB%9F%E5%B9%B2%E4%BA%86%E4%BB%80%E4%B9%88
https://dreampiggy.com/2022/11/07/%E6%8E%A2%E7%A7%98%E8%B6%8A%E6%9D%A5%E8%B6%8A%E5%A4%8D%E6%9D%82%E7%9A%84ImageIO%E6%A1%86%E6%9E%B6/

**Comment by iamsanketray** [comment]
Is this issue resolved?

**Comment by zhouyantongiosDev** [comment]
NO




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "SDWebImage/SDWebImage"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年1月11日(星期三) 晚上7:58
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [SDWebImage/SDWebImage] Memory usage increases when webp is used after 5.13.0 (Issue #3417)





 
Is this issue resolved?

**Comment by iamsanketray** [comment]
Could anyone confirm that downgrading to 5.12.6 will resolve this?

@dreampiggy I feel like the memory consumption displayed is correct as, I did get memory pressure warning.

**Comment by dreampiggy** [comment]
WebP decoding is not effected by SD 5.13's force decoding hack.
Only JPEG/HEIC get effected.

Check changlog of 5.14.3 and https://github.com/SDWebImage/SDWebImage/pull/3461, you can also try with 5.14.3 and test again.

**Comment by dreampiggy** [comment]
Or, a possible related PR in 5.13.0

See https://github.com/SDWebImage/SDWebImage/pull/3355, which use UIKit for force decode (to works for JPEG/HEIF YUV optimization)

Is this one does consume more RAM for WebP ? Than the old CoreGraphics force decode solution?

You can have a test by changeing SDImageCoderHelper.defaultDecodeSolution into CoreGraphis.

**Comment by dreampiggy** [comment]
@iamsanketray @kirou @zhouyantongiosDev  Any extra information ?

May wait for this issue, or may release 5.15.0

**Comment by iamsanketray** [comment]
@dreampiggy I can dive deep and check if you want me to. You will have to tell me what exactly to do. 

All I can tell you is that, I used PINRemote and it doesn't have the same memory issue.

**Comment by dreampiggy** [comment]
Do test in these cases:

1. Use [5.14.3](https://github.com/SDWebImage/SDWebImage/releases) on iOS 15+  to get a memory consumation report (using Instruments allocation template)
2. Use [5.14.3](https://github.com/SDWebImage/SDWebImage/releases) on iOS 14-  to get a memory consumation report (using Instruments allocation template)
3. Added the code in your Appdelegate `SDImageCodersHelper.defaultDecodeSolution = SDImageDecodeSolutionCoreGraphics`, then on iOS 15+  to get a memory consumation report (using Instruments allocation template)
4. Added the code in your Appdelegate `SDImageCodersHelper.defaultDecodeSolution = SDImageDecodeSolutionCoreGraphics`, then on iOS 14-  to get a memory consumation report (using Instruments allocation template)

**Comment by dreampiggy** [comment]
Save the `.trace` file and provided to me for investigation.

Or you can added a little Demo project, to let me to check

**Comment by iamsanketray** [comment]
`SDImageCoderHelper.defaultDecodeSolution = .coreGraphics` seems to solved the problem for me! Maybe we should add this somewhere in the docs.

**Comment by iamsanketray** [comment]
Could you tell me why this was fixed by changing the default decode colution?

**Comment by iamsanketray** [comment]
@dreampiggy Sorry for the late response!

`SDImageCoderHelper.defaultDecodeSolution = .coreGraphics` seems to fix the problem for me. Could you tell me why this fixes the problem? And, maybe we should add this in the document somewhere to help others.

**Comment by dreampiggy** [comment]
You can provide a simple demo for me to investigate the issue.

Here are code to only filter `JPEG/HEIF` format and use UIKit's decode solution, so I wonder why your **WebP** images effected by this.

**Comment by zhouyantongiosDev** [comment]
> @iamsanketray @kirou @zhouyantongiosDev Any extra information ?
> 
> May wait for this issue, or may release 5.15.0

Our phenomenon is that after replacement, sliding a large amount of webp images on the business page may cause memory issues and crashes. We resolved this issue after rolling back to version 5.13.2

**Comment by Lguanghui** [comment]
我这里也有同样的问题。同样的页面，从 5.12.5 升级到 5.13.x 之后，内存占用飙升。
![image](https://github.com/SDWebImage/SDWebImage/assets/59787459/39969d90-f162-48ab-ade6-41b7acc56fc6)
![image](https://github.com/SDWebImage/SDWebImage/assets/59787459/9fbc95f8-c06a-4569-88d5-c1d90a3931a1)
![Screen Shot 2023-06-29 at 11 45 39](https://github.com/SDWebImage/SDWebImage/assets/59787459/84d61fcb-b5ca-441a-830b-c5a2df8a344d)

**Comment by dreampiggy** [comment]
> Update the decode solution to allows CoreGraphics avoid using any UIKit method #3468

Don'y you see that `Other process RAM` increase when your `App RAM` decrease ?

Seems you don't understand what that `Force Decode` used for, see: https://dreampiggy.com/2022/11/07/%E6%8E%A2%E7%A7%98%E8%B6%8A%E6%9D%A5%E8%B6%8A%E5%A4%8D%E6%9D%82%E7%9A%84ImageIO%E6%A1%86%E6%9E%B6/

https://dreampiggy.com/2017/10/30/iOS%E5%B9%B3%E5%8F%B0%E5%9B%BE%E7%89%87%E7%BC%96%E8%A7%A3%E7%A0%81%E5%85%A5%E9%97%A8%E6%95%99%E7%A8%8B%EF%BC%88Image:IO%E7%AF%87%EF%BC%89/

**Comment by Lguanghui** [comment]
Using **SDImageAWebPCoder** on iOS 14+ solved my problem, but still have memory issue on iOS 13-. I have tried to set **defaultDecodeSolution** to **SDImageCoderDecodeSolutionCoreGraphics** on iOS 13-.

**Comment by Lguanghui** [comment]
> 
🙏

**Comment by Lguanghui** [comment]
> > Update the decode solution to allows CoreGraphics avoid using any UIKit method #3468
> 
> Don'y you see that `Other process RAM` increase when your `App RAM` decrease ?
> 
> Seems you don't understand what that `Force Decode` used for, see: https://dreampiggy.com/2022/11/07/%E6%8E%A2%E7%A7%98%E8%B6%8A%E6%9D%A5%E8%B6%8A%E5%A4%8D%E6%9D%82%E7%9A%84ImageIO%E6%A1%86%E6%9E%B6/
> 
> https://dreampiggy.com/2017/10/30/iOS%E5%B9%B3%E5%8F%B0%E5%9B%BE%E7%89%87%E7%BC%96%E8%A7%A3%E7%A0%81%E5%85%A5%E9%97%A8%E6%95%99%E7%A8%8B%EF%BC%88Image:IO%E7%AF%87%EF%BC%89/

🙏

**Comment by Lguanghui** [comment]
![image](https://github.com/SDWebImage/SDWebImage/assets/59787459/e16338ab-1100-49d0-ade1-997f73c92c2b)

**Comment by dreampiggy** [comment]
> seems these memory malloc for WebP decoding not released.
@Lguanghui Provide you demo ?

The `SDAnimatedImageView` will retain a `SDAniamtedImagePlayer`, which retains `SDAnimatedImage`, which retains `SDImageWebPCoder`, which retain the `WebPCodec`, which retinas the `WebPData`.

If the image view is still playing, it must be available. Even if you release the image view the RAM is still occupied ?

**Comment by dreampiggy** [comment]
Wait, your screenshot seems does not use the `SDAnimatedImageView`, but the normal UIImageView and UIImage ?

At that time, the buffer is retained by CGImage, when CGImage dealloc, it will call the release callback we provided, and call `WebPFree` on buffer.

You can check whether the RAM occupied decrease if you releaes the UIImage (by using like `UIImageView.image = nil`)

**Comment by Lguanghui** [comment]
> Wait, your screenshot seems does not use the `SDAnimatedImageView`, but the normal UIImageView and UIImage ?
> 
> At that time, the buffer is retained by CGImage, when CGImage dealloc, it will call the release callback (`CGDataProviderReleaseDataCallback`) we provided, and call `WebPFree` on buffer.
> 
> You can check whether the RAM occupied decrease if you releaes the UIImage (by using like `UIImageView.image = nil`)

Yes, the cell is using UIImageView and UIImage.  Both **using `UIImageView.image = nil`** and just going back to the previous page (which will **dealloc UIViewController, as well as the image cells**) have no effect on reducing memory.

- SDWebImage version: 5.16.0
- SDWebImageWebPCoder version: 0.12.0

my config:
<img width="2349" alt="image" src="https://github.com/SDWebImage/SDWebImage/assets/59787459/7b7efa75-aecc-421e-93d5-07a134bee2af">

You can get more information via this [video](https://drive.google.com/file/d/1IvDyNTHdcw_lSXb8b-tTHcwRVVHY9JeN/view?usp=sharing), which tested on an iOS 13 device.

**Comment by zhouyantongiosDev** [comment]
我们的现象是业务页面有大量wbep的图片就会出现卡的问题,甚至闪退. 直到回退到5.13.2版本并且上线一年,目前看着啥问题。

**Comment by Lguanghui** [comment]
我们项目里默认全部在 CDN 上转成 webp 下载，所以问题也挺严重😂

> 我们的现象是业务页面有大量wbep的图片就会出现卡的问题,甚至闪退. 直到回退到5.13.2版本并且上线一年,目前看着啥问题。

**Comment by zhouyantongiosDev** [comment]
> 我们有过评估和测试,从压缩,解压缩,图片质量等角度评估并不是所有图片都适合webp,所以我们存在好好几种类型的图片

**Comment by dreampiggy** [comment]
Provide a reproducable demo about this performance behavior...Currently I can not reproduce using SDWebImageWebPCoder demo between versions...

And @Lguanghui Say that `This behavior is about SDWebImage version change`, not `SDWebImageWebPCoder` version change, which is strange...

**Comment by dreampiggy** [comment]
If you two want more details, contact me via my email lizhuoli1126@126.com or even some IM (I'll reply to you the wechat-account in email)

**Comment by Lguanghui** [comment]
> If you two want more details, contact me via my email [lizhuoli1126@126.com](mailto:lizhuoli1126@126.com) or even some IM (I'll reply to you the wechat-account in email)

Wait for a while, I will send a reproducable demo to you via your email.

**Comment by dreampiggy** [comment]
Effected: 5.13.1 ~ 5.16.0

Should be totally solved by #3559 and WebPCoder https://github.com/SDWebImage/SDWebImageWebPCoder/pull/77

If you need low version back-port, I can consider a 5.16.1 version by only revert the changes cause the issue.
